supervisor/homeassistant/websocket.py

```diff
@@ -22,61 +26,112 @@ class WSClient:
+        _LOGGER.debug("Sending: %s", message)
+        try:
+            await self._client.send_json(message)
+        except ConnectionError as err:
+            raise HomeAssistantWSConnectionError(err) from err
 
-    async def async_send_command(self, message: dict[str, Any]):
-        """Send a websocket command."""
-        async with self._lock:
-            self.message_id += 1
-            message["id"] = self.message_id
+        try:
+            return await self._futures[message["id"]]
+        finally:
+            self._futures.pop(message["id"])
 
-            _LOGGER.debug("Sending: %s", message)
-            try:
-                await self.client.send_json(message)
-            except ConnectionError:
-                return
+    async def start_listener(self) -> None:
+        """Start listening to the websocket."""
+        if not self.connected:
+            raise HomeAssistantWSConnectionError("Not connected when start listening")
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
+
+        except HomeAssistantWSError as err:
+            _LOGGER.warning(err)
+
+        finally:
+            await self.close()
+
+    async def _receive_json(self) -> None:
+        """Receive json."""
+        msg = await self._client.receive()
+        _LOGGER.debug("Received: %s", msg)
 
-            _LOGGER.debug("Received: %s", response)
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
+
+        if msg.type != WSMsgType.TEXT:
+            raise HomeAssistantWSError(f"Received non-Text message: {msg.type}")
+
+        try:
+            data = msg.json()
+        except ValueError as err:
+            raise HomeAssistantWSError(f"Received invalid JSON - {msg}") from err
 
-            if response["success"]:
-                return response["result"]
+        if data["type"] == "result":
+            if (future := self._futures.get(data["id"])) is None:
+                return
+
+            if data["success"]:
+                future.set_result(data["result"])
+                return
 
-            raise HomeAssistantWSError(response)
+            _LOGGER.error("Unsuccessful websocket message - %s", data)
 
     @classmethod
     async def connect_with_auth(
-        cls, session: aiohttp.ClientSession, url: str, token: str
+        cls, session: aiohttp.ClientSession, loop, url: str, token: str
     ) -> "WSClient":
         """Create an authenticated websocket client."""
         try:
-            client = await session.ws_connect(url, ssl=False)
+            client = await session.ws_connect(url, timeout=500)
```

Comment:
So this sets `Timeout for websocket to close` to 8.3 minutes, any specific reason for that number? Do we have a guarantee to have a message within this time? Should we set it to `None`?

Reply:
This is outdated